File: binary_trees/tree_sum/tree_sum.py
from typing import Optional
class Node:
    def __init__(self, val):
        self.val = val
        self. left = None
        self.right = None

    def __str__(self):
        return f"value: {self.val}, left: {self.left}, right: {self.right}"

#      1
#     / \
#    2   3
#   /\   /
#  4  5 6
a = Node(1)
b = Node(2)
c = Node(3)
d = Node(4)
e = Node(5)
f = Node(6)

a.left = b
a.right = c
b.left = d
b.right = e
c.left = f

def tree_sum_bfs(root: Optional[Node]) -> int:
    count = 0
    queue = [root]

    while(len(queue) > 0):
        current = queue.pop(0)
        count += current.val

        if current.left != None:
            queue.append(current.left)
        if current.right != None:
            queue.append(current.right)
    
    return count

def tree_sum_dfs(root: Optional[Node]) -> int:
    count = 0
    stack = [root]

    while(len(stack) > 0):
        current = stack.pop()
        count += current.val

        if current.left != None:
            stack.append(current.left)
        if current.right != None:
            stack.append(current.right)
    
    return count


def tree_sum_dfs_recursion(root: Optional[Node]) -> int:
    count = 0
    def recursion(root):
        nonlocal count
        if root != None:
            count += root.val
            if root.left != None:
                recursion(root.left)
            if root.right != None:
                recursion(root.right)

    recursion(root)
        
    return count

print(tree_sum_bfs(a))
print(tree_sum_dfs(a))
print(tree_sum_dfs_recursion(a))
print(tree_sum_bfs(b))
print(tree_sum_dfs(b))
print(tree_sum_dfs_recursion(b))

File: binary_trees/tree_sum/tree_sum.js
class Node {
    constructor(val) {
        this.val = val;
        this.left = null;
        this.right = null;
    };
}

const a = new Node(1);
const b = new Node(2);
const c = new Node(3);
const d = new Node(4);
const e = new Node(5);
const f = new Node(6);

//      a
//     / \
//    b   c
//   /\   /
//  d  e f
a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.left = f;

function tree_sum_bfs(root) {
    let count = 0;
    let queue = [root];

    while(queue.length > 0) {
        let current = queue.shift();
        if(current) {
            count += current.val;
        }
        if(current.left) {
            queue.push(current.left);
        }
        if(current.right) {
            queue.push(current.right);
        }
    }

    return count;
}

function tree_sum_dfs(root) {
    let count = 0;
    let stack = [root];

    while(stack.length > 0) {
        let current = stack.pop();

        if(current) {
            count += current.val;
        }
        if(current.left) {
            stack.push(current.left);
        }
        if(current.right) {
            stack.push(current.right);

        }
    }

    return count;
}

let count = 0;
function tree_sum_dfs_recursion(root) {
    if(root) {
        count += root.val;
    }
    if(root.left) {
        tree_sum_dfs_recursion(root.left);
    }
    if(root.right) {
        tree_sum_dfs_recursion(root.right);
    }

    return count;
}

console.log(tree_sum_bfs(a));
console.log(tree_sum_dfs(a));
console.log(tree_sum_dfs_recursion(a));